### PR TITLE
python3-jetson-stats: update to 4.1.5

### PR DESCRIPTION
**Pull request by ichergui**
IMG-2511
IMG-2512
IMG-2513
IMG-2514
IMG-2515
IMG-2516
IMG-2517

### CODE REVIEW

**Comment by ichergui** [comment]
* Tests were done on machine  `Jetson Xavier AGX`  with image `demo-image-full`
```
root@jetson-agx-xavier-devkit:~# systemctl status jtop 
● jtop.service - jtop service
     Loaded: loaded (/lib/systemd/system/jtop.service; enabled; preset: enabled)
     Active: active (running) since Thu 2023-01-26 08:12:24 UTC; 3 weeks 2 days ago
   Main PID: 438 (python3)
      Tasks: 4 (limit: 37120)
     Memory: 31.5M
     CGroup: /system.slice/jtop.service
             ├─438 python3 /usr/bin/jtop --force
             ├─648 python3 /usr/bin/jtop --force
             └─663 python3 /usr/bin/jtop --force

Jan 26 08:12:28 jetson-agx-xavier-devkit jtop[438]: [INFO] jtop.service - Running on Python: 3.11.1
Jan 26 08:12:28 jetson-agx-xavier-devkit jtop[438]: [INFO] jtop.core.common - fan loaded on /sys/devices/platform
Jan 26 08:12:28 jetson-agx-xavier-devkit jtop[438]: [INFO] jtop.core.common - jetson_clocks loaded on /usr/bin/jetson_clocks
Jan 26 08:12:28 jetson-agx-xavier-devkit jtop[438]: [INFO] jtop.core.engine - Special Engine group found: [dlaX]
Jan 26 08:12:28 jetson-agx-xavier-devkit jtop[438]: [INFO] jtop.core.engine - Special Engine group found: [pvaX]
Jan 26 08:12:28 jetson-agx-xavier-devkit jtop[438]: [INFO] jtop.core.engine - Engines found: [APE CVNAS DLA0 DLA1 NVDEC NVENC NVJPG PVA0 PVA1 SE VIC]
Jan 26 08:12:28 jetson-agx-xavier-devkit jtop[438]: [INFO] jtop.core.common - tegrastats loaded on /usr/bin/tegrastats
Jan 26 08:12:28 jetson-agx-xavier-devkit jtop[438]: [INFO] jtop.__main__ - jetson_stats 4.1.5 - server loaded
Jan 26 08:12:30 jetson-agx-xavier-devkit jtop[663]: [INFO] jtop.core.jetson_clocks - Store jetson_clocks configuration in /usr/local/jtop/l4t_dfs.conf
Jan 26 08:12:30 jetson-agx-xavier-devkit jtop[663]: [INFO] jtop.core.fan - Mode set system status=False
root@jetson-agx-xavier-devkit:~#
```

**Comment by ichergui** [comment]
**Note**: I need to fix getting version of `CUDA`, `cuDNN`, `TRT`, `OpenCV` ...
If anyone can help, that would be much appreciated.

**Comment by dwalkes** [comment]
@ichergui regarding
>  I need to fix getting versions of CUDA, cuDNN, TRT, OpenCV ...

I can see the logic at https://github.com/rbonghi/jetson_stats/blob/bf49aab10e1ab1e34722c77bf21383a0ce9d9afa/jtop/core/jetson_libraries.py#L34 and I can reproduce with:
```
root@jetson-xavier-nx-devkit:~# python3 -m jtop.core.jetson_libraries
CUDA not installed!
OPENCV not Installed!
cuDNN:
TensorRT:
VPI:
Vulkan: 1.3.239
```
What's the correct way to get each of these for OE4T builds?  Should we get package versions for each and inject into a file we add to the recipe for jetson-stats?

**Comment by madisongh** [comment]
> What's the correct way to get each of these for OE4T builds? Should we get package versions for each and inject into a file we add to the recipe for jetson-stats?

For CUDA: I can't find where that `/usr/local/cuda/version.txt` file comes from that it's looking for. The fallback method, using `nvcc`, only works if you have the CUDA compiler installed.
For OpenCV: the `opencv_version` program is installed with the `opencv-apps` package, but that also pulls in a bunch of other programs.  Seems like overkill.
For the others: it uses `dpkg -l`, which obviously only works if you're using deb packaging and have `package-management` in your image features.

I don't see a nice and easy way to translate this functionality into OE builds.  You don't want to introduce new dependencies just for this feature, and you can't know at build time which of the other packages, if any, are going to be installed.  Once the rootfs is created for the image, you could have a postprocessing command to populate package version information into a file in the rootfs somewhere, and modify the function here to read that, but that's not something you could do from a recipe.  Perhaps you could add a bbclass to insert such a function into `ROOTFS_POSTPROCESS_COMMAND` and let the user decide whether or not it's important enough to them to add that to their image recipes.  I would expect, though, that developers that would use this tool would also be doing their own builds and would have other ways to know which versions of these packages are installed.   

IMHO, this seems like a bunch of work for very little actual value.

**Comment by ichergui** [comment]
> @ichergui regarding
> 
> > I need to fix getting versions of CUDA, cuDNN, TRT, OpenCV ...
> 
> I can see the logic at https://github.com/rbonghi/jetson_stats/blob/bf49aab10e1ab1e34722c77bf21383a0ce9d9afa/jtop/core/jetson_libraries.py#L34 and I can reproduce with:
> 
> ```
> root@jetson-xavier-nx-devkit:~# python3 -m jtop.core.jetson_libraries
> CUDA not installed!
> OPENCV not Installed!
> cuDNN:
> TensorRT:
> VPI:
> Vulkan: 1.3.239
> ```
> 
> What's the correct way to get each of these for OE4T builds? Should we get package versions for each and inject into a file we add to the recipe for jetson-stats?

`CUDA` I can propose a solution for that, there is 2 alternative : 
1- Using cuda samples .e.g. deviceQuery and parse the result and get the CUDA version 
2- Installing `nvcc` and the job will be done easy 

`OpenCV` the runtime dependency is missing. I will test that and will be back to you 

My concern is more about `TRT`, `cuDNN` and `VPI`

**Comment by ichergui** [comment]
> > What's the correct way to get each of these for OE4T builds? Should we get package versions for each and inject into a file we add to the recipe for jetson-stats?
> 
> For CUDA: I can't find where that `/usr/local/cuda/version.txt` file comes from that it's looking for. The fallback method, using `nvcc`, only works if you have the CUDA compiler installed. For OpenCV: the `opencv_version` program is installed with the `opencv-apps` package, but that also pulls in a bunch of other programs. Seems like overkill. For the others: it uses `dpkg -l`, which obviously only works if you're using deb packaging and have `package-management` in your image features.
> 
> I don't see a nice and easy way to translate this functionality into OE builds. You don't want to introduce new dependencies just for this feature, and you can't know at build time which of the other packages, if any, are going to be installed. Once the rootfs is created for the image, you could have a postprocessing command to populate package version information into a file in the rootfs somewhere, and modify the function here to read that, but that's not something you could do from a recipe. Perhaps you could add a bbclass to insert such a function into `ROOTFS_POSTPROCESS_COMMAND` and let the user decide whether or not it's important enough to them to add that to their image recipes. I would expect, though, that developers that would use this tool would also be doing their own builds and would have other ways to know which versions of these packages are installed.
> 
> IMHO, this seems like a bunch of work for very little actual value.

Yes, agree on the logic and the conclusion.

**Comment by ichergui** [comment]
* Here is the result after adding some updates 
```
root@jetson-agx-xavier-devkit:~# python3 -m jtop.core.jetson_libraries
CUDA: 11.4.315
OPENCV: 4.6.0 - with CUDA: YES
cuDNN: 
TensorRT: 
VPI: 
Vulkan: 1.3.239
root@jetson-agx-xavier-devkit:~# 

```

**Comment by ichergui** [comment]
@madisongh @dwalkes Let me know guys if you should stop here.
 I completely agree with you @madisongh. This seems like a bunch of work for very little actual value.
We can think about improving things later.

Please share your thoughts.

**Comment by dwalkes** [comment]
Would this work reliably?
```
re.search('libcudnn.so.(\d+.\d+.\d+)',str(list(Path('/usr/lib/').glob("libcudnn.so.*.*"))[0])).group(1)
'8.6.0'
```

**Comment by ichergui** [comment]
> Would this work reliably?
> 
> ```
> re.search('libcudnn.so.(\d+.\d+.\d+)',str(list(Path('/usr/lib/').glob("libcudnn.so.*.*"))[0])).group(1)
> '8.6.0'
> ```

I think that works fine 
I get try it. Thanks @dwalkes

File: recipes-devtools/python/python3-jetson-stats_4.1.5.bb [hunk 1]
```diff
@@ -41,11 +44,16 @@ RDEPENDS:${PN} += " \
     python3-json \
     python3-multiprocessing \
     python3-setuptools \
+    python3-smbus2 \
+    python3-distro \
     tegra-nvpmodel \
     tegra-tools-jetson-clocks \
     tegra-tools-tegrastats \
+    nv-tegra-release \
+    cuda-nvcc \
```
[the remark this answers is not in the document]
Comment (ichergui) [reply recipes-devtools/python/python3-jetson-stats_4.1.5.bb:53]: Getting the `nvcc` installed into the image facilitate retrieve the `CUDA` version
Do you prefer to run the cuda sample like `deviceQuery` then parse the result to get the `CUDA` version  ?
I will remove `opencv` from the runtime dependency list ?

[the remark this answers is not in the document]
Comment (madisongh) [reply recipes-devtools/python/python3-jetson-stats_4.1.5.bb:53]: I'd just remove both, and maybe put a comment in the recipe to tell the user to add the relevant packages to their builds if they care about having this tool report the version.

**Comment by ichergui** [comment]
* Tests were done on machine `Jetson Xavier AGX` with different images  
* Add the following line into `conf/local.conf`
```
CORE_IMAGE_BASE_INSTALL:append = " python3-jetson-stats"
```

* Result using `demo-image-full` image 
```
root@jetson-agx-xavier-devkit:~# python3 -m jtop.core.jetson_libraries
CUDA: 11.4
OPENCV: not Installed!
TensorRT: 8.5.2
cuDNN: 8.6.0
VPI: 2.2.4
Vulkan: 1.3.239
root@jetson-agx-xavier-devkit:~# 
```

* Result using `demo-image-base` image 
```
root@jetson-agx-xavier-devkit:~# python3 -m jtop.core.jetson_libraries
CUDA: 11.4
OPENCV: not Installed!
TensorRT: not Installed!
cuDNN: not Installed!
VPI: not Installed!
root@jetson-agx-xavier-devkit:~# 
```

**Pull request merged**